lib/Sema/CSDiagnostics.cpp

```diff
@@ -3057,11 +3057,17 @@ bool ContextualFailure::diagnoseConversionToBool() const {
   auto *anchor = castToExpr(getAnchor());
   // Check for "=" converting to Bool.  The user probably meant ==.
   if (auto *AE = dyn_cast<AssignExpr>(anchor->getValueProvidingExpr())) {
-    emitDiagnosticAt(AE->getEqualLoc(), diag::use_of_equal_instead_of_equality)
-        .fixItReplace(AE->getEqualLoc(), "==")
-        .highlight(AE->getDest()->getLoc())
-        .highlight(AE->getSrc()->getLoc());
-    return true;
+    if (AE->getDest()->getKind() == ExprKind::Error && !findParentExpr(AE)) {
+      // Already diagnosed as user probably meant a pattern binding
+      return true;
```

Comment:
This makes me wonder whether the approach we currently take during pre-checking is not the right one. We didn't have placeholders or anything like that before when this was added but I think instead of error expressions we could use a "placeholder" AST (i.e. EditorPlaceholderExpr or create a new one) node with a `PlaceholderType` in it which would mean that we won't end up producing unrelated fixes and diagnostics like this one if cases were something couldn't be found in scope...

Reply:
I can see the use of that sort of design, but I think the difficulty comes in deciding how far you bubble up the placeholders and how. Like here, you don't want to lose the whole AssignExpr, because the Src is still useful, so I think this sort of code that looks down at the Dest for either Error or theoretically `PlaceholderType` would look pretty much the same.

Reply:
Only destination is going to get a placeholder type and propagation is already handled by the solver since that’s the “hole” type.

Reply:
In more general terms any declaration we cannot find results for would be a placeholder. We do this for members already with DefineBasedOnUse fix.

Reply:
I can see in general it would be better to replace with placeholders instead of error expressions.

But in this particular case (the destination of an assignment) the code impact turns out to be fairly spread out, because of the transformation needed. 

We'd discover during pre-check that instead of a `SyntacticElementTarget` that is an expression, it ought to be a pattern binding, and then potentially continue the type check with the rest of the expression as a `CTP_Initialization`. And the pre-check code is happening in static members, so we don't have a `ConstraintSystem` to add a fix to yet, at this point.

I'm hesitant to make that big of a change to the structure of the code that invokes type checking for this without being fairly sure that it's something you'd approve.

Alternatives:
1) What I take about in this comment: change `typeCheckTarget` (and related bits) to make this sort of change of target possible.
2) Just make `preCheckTarget` fail when we diagnose this, so we get the error and skip type checking the initialization expression. (Same results as current PR but some further diagnoses of the rest of expressions go away, and this `// already diagnosed` bit of code no longer needed.)
3) This as-is.
What do you think?